File: bcs-services/bcs-storage/storage/actions/lib/manager.go
/*
 * Tencent is pleased to support the open source community by making Blueking Container Service available.
 * Copyright (C) 2019 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package lib

import (
	"fmt"
	"sync"
	"time"

	"github.com/Tencent/bk-bcs/bcs-common/common/blog"
)

// Manager manager
type Manager struct {
	total   int64
	avgTime int64

	subTotal   int64
	subAvgTime int64

	lt10  int64
	lt50  int64
	lt100 int64
	lt200 int64
	gt200 int64

	sep time.Duration
	sync.Mutex

	title string
}

// NewManager create manager
func NewManager(sep int64, title string) *Manager {
	return &Manager{
		sep:   time.Duration(sep),
		title: title,
	}
}

// Start start
func (m *Manager) Start() {
	for {
		m.Flush()
		time.Sleep(m.sep * time.Second)
	}
}

// Add add
func (m *Manager) Add(t time.Duration) {
	_t := int64(t)
	m.Lock()
	defer m.Unlock()
	m.subAvgTime = (m.subAvgTime*m.subTotal + _t) / (m.subTotal + 1)
	m.subTotal++

	if t < 10*time.Millisecond {
		m.lt10++
	}
	if t < 50*time.Millisecond {
		m.lt50++
	}
	if t < 100*time.Millisecond {
		m.lt100++
	}
	if t < 200*time.Millisecond {
		m.lt200++
	}
	if t >= 200*time.Millisecond {
		m.gt200++
	}
}

// Flush manager do flush
func (m *Manager) Flush() {
	m.Lock()
	defer m.Unlock()
	if m.total+m.subTotal == 0 {
		m.avgTime = 0
	} else {
		m.avgTime = (m.avgTime*m.total + m.subAvgTime*m.subTotal) / (m.total + m.subTotal)
	}
	m.total += m.subTotal

	blog.Infof(
		"\n%s | %s"+
			"\n%s | %-20s %-20s %-20s %-20s [per %d second]"+
			"\n%s | %-20d %-20f %-20d %-20f"+
			"\n%s |"+
			"\n%s | %-20s %-20s %-20s %-20s %-20s"+
			"\n%s | %-20s %-20s %-20s %-20s %-20s"+
			"\n%s | ---------------------------------------------------------------------------------------------------\n",
		m.title, time.Now().String(),
		m.title, "TOTAL", "AVG_TIME(ms)", "SUB_TOTAL", "SUB_AVG_TIME(ms)", m.sep,
		m.title, m.total, float64(m.avgTime)/1e6, m.subTotal, float64(m.subAvgTime)/1e6,
		m.title,
		m.title, "<10ms", "<50ms", "<100ms", "<200ms", ">=200ms",
		m.title, fmt.Sprintf("%.2f%%", float64(m.lt10)/float64(m.total)*100),
		fmt.Sprintf("%.2f%%", float64(m.lt50)/float64(m.total)*100),
		fmt.Sprintf("%.2f%%", float64(m.lt100)/float64(m.total)*100),
		fmt.Sprintf("%.2f%%", float64(m.lt200)/float64(m.total)*100),
		fmt.Sprintf("%.2f%%", float64(m.gt200)/float64(m.total)*100),
		m.title)

	m.subTotal = 0
	m.subAvgTime = 0
}

File: bcs-ui/backend/templatesets/release/manager.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
import functools
from typing import List, Tuple

from backend.resources.utils.kube_client import CoreDynamicClient
from backend.utils.async_run import async_run
from backend.utils.basic import getitems

from .. import models


class AppReleaseManager:
    """
    向上层提供模板集 release 的 CRUD 操作，职责如下
    - 管理 AppRelease 表，完成 release 数据的 CRUD
    - 通过 ReleaseResourceManager 管理 release 中的 resource 变更
    - 一个 release 至少包含一个 resource, 一个 resource 对应一个 ReleaseResourceManager
    """

    def __init__(self, dynamic_client: CoreDynamicClient):
        self.dynamic_client = dynamic_client

    def update_or_create(self, operator: str, release_data: models.AppReleaseData) -> Tuple[models.AppRelease, bool]:
        app_release, created = models.AppRelease.objects.update_or_create(
            name=release_data.name,
            cluster_id=release_data.cluster_id,
            namespace=release_data.namespace,
            defaults={
                "template_id": release_data.template_id,
                "project_id": release_data.project_id,
                "creator": operator,
                "updator": operator,
            },
        )

        try:
            self._deploy(operator, app_release.id, release_data.resource_list)
        except Exception as e:
            app_release.update_status(models.ReleaseStatus.FAILED.value, str(e))
        else:
            app_release.update_status(models.ReleaseStatus.DEPLOYED.value)

        return app_release, created

    def _deploy(self, operator: str, app_release_id: int, resource_list: List[models.ResourceData]):
        res_mgr = ReleaseResourceManager(self.dynamic_client, app_release_id)
        tasks = [functools.partial(res_mgr.update_or_create, operator, resource) for resource in resource_list]
        async_run(tasks)

    def delete(self, operator: str, app_release_id: int):
        self._delete(operator, app_release_id)
        models.AppRelease.objects.filter(id=app_release_id).delete()

    def _delete(self, operator: str, app_release_id: int):
        res_mgr = ReleaseResourceManager(self.dynamic_client, app_release_id)
        tasks = [
            functools.partial(res_mgr.delete, operator, resource_inst.id)
            for resource_inst in models.ResourceInstance.objects.filter(app_release_id=app_release_id)
        ]
        async_run(tasks)


class ReleaseResourceManager:
    """
    提供对 AppRelease 中 resource 的 CRUD 操作
    - 管理 ResourceInstance 表，完成 resource 数据的 CRUD
    - 通过与集群的实际对接，管理 resource 的运行状态
    """

    def __init__(self, dynamic_client: CoreDynamicClient, app_release_id: int):
        self.dynamic_client = dynamic_client
        self.app_release_id = app_release_id

    def update_or_create(self, operator: str, resource: models.ResourceData) -> Tuple[models.ResourceInstance, bool]:
        api = self._get_api(kind=resource.kind, api_version=getitems(resource.manifest, 'apiVersion'))
        api.update_or_create(
            body=resource.manifest,
            name=resource.name,
            namespace=getitems(resource.manifest, 'metadata.namespace'),
            update_method="replace",
        )

        # https://youngminz.netlify.app/posts/get-or-create-deadlock
        # 并发条件下update_or_create 触发 deadlock found when trying to get lock; try restarting transaction
        kwargs = {'app_release_id': self.app_release_id, 'kind': resource.kind, 'name': resource.name}
        defaults = {
            'manifest': resource.manifest,
            'version': resource.version,
            'revision': resource.revision,
            'updator': operator,
            'creator': operator,
        }
        try:
            resource_inst = models.ResourceInstance.objects.get(**kwargs)
        except models.ResourceInstance.DoesNotExist:
            kwargs.update(defaults)
            return models.ResourceInstance.objects.create(**kwargs), True
        else:
            resource_inst.__dict__.update(defaults)
            resource_inst.save()
            return resource_inst, False

    def edit(self, operator: str, resource: models.ResourceData):
        """
        直接edit线上资源(类似kubectl edit)，不做manifest和version等信息的保存
        """
        api = self._get_api(kind=resource.kind, api_version=getitems(resource.manifest, 'apiVersion'))
        api.update_or_create(
            body=resource.manifest,
            name=resource.name,
            namespace=getitems(resource.manifest, 'metadata.namespace'),
            update_method="replace",
        )
        models.ResourceInstance.objects.filter(
            app_release_id=self.app_release_id, name=resource.name, namespace=resource.namespace, kind=resource.kind
        ).update(edited=True, updator=operator)

    def delete(self, operator: str, resource_inst_id: int):
        resource_inst = models.ResourceInstance.objects.get(id=resource_inst_id)
        api = self._get_api(resource_inst.kind, getitems(resource_inst.manifest, 'apiVersion'))
        api.delete_ignore_nonexistent(
            name=resource_inst.name, namespace=getitems(resource_inst.manifest, 'metadata.namespace')
        )
        return resource_inst.delete()

    def _get_api(self, kind, api_version):
        if api_version:
            return self.dynamic_client.resources.get(kind=kind, api_version=api_version)
        return self.dynamic_client.get_preferred_resource(kind=kind)
